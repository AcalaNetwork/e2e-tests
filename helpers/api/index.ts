import { ApiPromise } from '@polkadot/api'

export const xtokens = {
  relaychainV2: (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            id: acc,
          },
        },
      },
    },
  }),
  parachainV2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  parachainAccountId20V2: (paraId: number) => (acc: any) => ({
    V1: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountKey20: {
              network: 'Any',
              key: acc,
            },
          },
        ],
      },
    },
  }),
  parachainV3: (paraId: number) => (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X2: [
          { Parachain: paraId },
          {
            AccountId32: {
              network: 'Any',
              id: acc,
            },
          },
        ],
      },
    },
  }),
  transfer:
    (token: any, amount: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transfer(token, amount, dest(acc), weight),
  transferMulticurrencies:
    (token: any, amount: any, feeToken: any, feeAmount: any, dest: (dest: any) => any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      api.tx.xTokens.transferMulticurrencies(
        [
          [token, amount],
          [feeToken, feeAmount],
        ],
        1,
        dest(acc),
        'Unlimited'
      ),
}

export const xcmPallet = {
  parachainV2: (parents: number, paraId: number) => ({
    V1: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  relaychainV3: (acc: any) => ({
    V3: {
      parents: 1,
      interior: {
        X1: {
          AccountId32: {
            network: 'Any',
            id: acc,
          },
        },
      },
    },
  }),
  parachainV3: (parents: number, paraId: any) => ({
    V3: {
      parents,
      interior: {
        X1: { Parachain: paraId },
      },
    },
  }),
  limitedReserveTransferAssetsV2:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V1: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  network: 'Any',
                  id: acc,
                },
              },
            },
          },
        },
        {
          V1: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
  limitedReserveTransferAssetsV3:
    (token: any, amount: any, dest: any) =>
    ({ api }: { api: ApiPromise }, acc: any) =>
      (api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
        dest,
        {
          V3: {
            parents: 0,
            interior: {
              X1: {
                AccountId32: {
                  id: acc,
                },
              },
            },
          },
        },
        {
          V3: [
            {
              id: token,
              fun: { Fungible: amount },
            },
          ],
        },
        0,
        'Unlimited'
      ),
}

export const tx = {
  xtokens,
  xcmPallet,
}

export const query = {
  balances: ({ api }: { api: ApiPromise }, address: string) => api.query.system.account(address),
  tokens:
    (token: any) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.tokens.accounts(address, token),
  assets:
    (token: number | bigint) =>
    ({ api }: { api: ApiPromise }, address: string) =>
      api.query.assets.account(token, address),
  evm:
    (contract: string, slot: string) =>
    ({ api }: { api: ApiPromise }, _address: string) =>
      api.query.evm.accountStorages(contract, slot),
}
