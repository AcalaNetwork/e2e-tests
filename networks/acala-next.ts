import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
