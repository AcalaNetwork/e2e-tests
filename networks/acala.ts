import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
