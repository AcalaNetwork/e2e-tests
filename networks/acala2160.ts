import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
