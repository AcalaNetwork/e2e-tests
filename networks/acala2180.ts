import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
