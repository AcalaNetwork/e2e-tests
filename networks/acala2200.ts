import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
