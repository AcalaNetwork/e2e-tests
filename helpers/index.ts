import { ApiPromise } from '@polkadot/api'
import { Codec } from '@polkadot/types/types'
import { expect } from 'vitest'

type CodecOrArray = Codec | Codec[]

const processCodecOrArray = (codec: CodecOrArray, fn: (c: Codec) => any) =>
  Array.isArray(codec) ? codec.map(fn) : fn(codec)

const toHuman = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toHuman?.() ?? c)
const toHex = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toHex?.() ?? c)
const toJson = (codec: CodecOrArray) => processCodecOrArray(codec, (c) => c?.toJSON?.() ?? c)

export type EventFilter = string | { method: string; section: string }

export type RedactOptions = {
  number?: boolean | number // precision
  hash?: boolean // 32 byte hex
  hex?: boolean // any hex with 0x prefix
  address?: boolean // base58 address
}

export class Checker {
  readonly #value: any
  readonly #pipeline: Array<(value: any) => any> = []

  #format: 'human' | 'hex' | 'json' = 'json'
  #message: string | undefined
  #redactOptions: RedactOptions | undefined

  constructor(value: any, message?: string) {
    this.#value = value
    this.#message = message
  }

  toHuman() {
    this.#format = 'human'
    return this
  }

  toHex() {
    this.#format = 'hex'
    return this
  }

  toJson() {
    this.#format = 'json'
    return this
  }

  message(message: string) {
    this.#message = message
    return this
  }

  filterEvents(...filters: EventFilter[]) {
    this.toHuman()
    this.#pipeline.push((value) => {
      let data = value.map(({ event: { index: _, ...event } }: any) => event)
      if (filters.length > 0) {
        data = data.filter((evt: any) => {
          return filters.some((filter) => {
            if (typeof filter === 'string') {
              return evt.section === filter
            } else if ('method' in filter) {
              const { section, method } = filter
              return evt.section === section && evt.method === method
            }
          })
        })
      }
      return data
    })
    return this
  }

  redact(options: RedactOptions = { number: 2, hash: true }) {
    this.#redactOptions = {
      ...this.#redactOptions,
      ...options,
    }
    return this
  }

  #redact(value: any) {
    if (!this.#redactOptions) {
      return value
    }

    const redactNumber = this.#redactOptions.number === true || typeof this.#redactOptions.number === 'number'
    const precision = redactNumber
      ? typeof this.#redactOptions.number === 'number'
        ? this.#redactOptions.number
        : 0
      : 0
    const redactHash = this.#redactOptions.hash === true
    const redactHex = this.#redactOptions.hex === true
    const redactAddress = this.#redactOptions.address === true

    const processNumber = (value: number) => {
      if (precision > 0) {
        const rounded = parseFloat(value.toPrecision(precision))
        if (rounded === value) {
          return rounded
        }
        return `(rounded ${rounded})`
      }
      return '(number)'
    }

    const process = (obj: any): any => {
      if (obj == null) {
        return obj
      }
      if (Array.isArray(obj)) {
        return obj.map(process)
      }
      if (redactNumber && typeof obj === 'number') {
        return processNumber(obj)
      }
      if (typeof obj === 'string') {
        if (redactNumber && obj.match(/0x000000[0-9a-f]{26}/)) {
          // this is very likely u128 encoded in hex
          const num = parseInt(obj)
          return processNumber(num)
        }
        if (redactHash && obj.match(/0x[0-9a-f]{64}/)) {
          return '(hash)'
        }
        if (redactHex && obj.match(/0x[0-9a-f]+/)) {
          return '(hex)'
        }
        if (redactAddress && obj.match(/^[1-9A-HJ-NP-Za-km-z]{46,48}$/)) {
          return '(address)'
        }
        if (redactNumber && obj.match(/^-?[\d,]+$/)) {
          const num = parseInt(obj.replace(/,/g, ''))
          return processNumber(num)
        }
        return obj
      }
      if (typeof obj === 'object') {
        return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, process(v)]))
      }
      return obj
    }

    return process(value)
  }

  map(fn: (value: any) => any) {
    this.#pipeline.push(fn)
    return this
  }

  pipe(fn?: (value: Checker) => Checker) {
    return fn ? fn(this) : this
  }

  async value() {
    let value = await this.#value

    switch (this.#format) {
      case 'human':
        value = toHuman(value)
        break
      case 'hex':
        value = toHex(value)
        break
      case 'json':
        value = toJson(value)
        break
    }

    for (const fn of this.#pipeline) {
      value = await fn(value)
    }

    value = this.#redact(value)

    return value
  }

  async toMatchSnapshot(msg?: string) {
    return expect(await this.value()).toMatchSnapshot(msg ?? this.#message)
  }
}

export const check = (value: any, msg?: string) => {
  if (value instanceof Checker) {
    if (msg) {
      return value.message(msg)
    }
    return value
  }
  return new Checker(value, msg)
}

type Api = { api: ApiPromise }

export const checkEvents = ({ events }: { events: Promise<Codec[] | Codec> }, ...filters: EventFilter[]) =>
  check(events, 'events')
    .filterEvents(...filters)
    .redact()

export const checkSystemEvents = ({ api }: Api, ...filters: EventFilter[]) =>
  check(api.query.system.events(), 'system events')
    .filterEvents(...filters)
    .redact()

export const checkUmp = ({ api }: Api) =>
  check(api.query.parachainSystem.upwardMessages(), 'ump').map((value) =>
    api.createType('Vec<XcmVersionedXcm>', value).toJSON()
  )

export const checkHrmp = ({ api }: Api) =>
  check(api.query.parachainSystem.hrmpOutboundMessages(), 'hrmp').map((value) =>
    (value as any[]).map(({ recipient, data }) => ({
      data: api.createType('(XcmpMessageFormat, XcmVersionedXcm)', data),
      recipient,
    }))
  )
