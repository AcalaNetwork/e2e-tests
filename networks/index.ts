import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
