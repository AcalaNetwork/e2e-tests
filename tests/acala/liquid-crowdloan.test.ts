import { beforeEach, describe, it } from 'vitest'
import { check, checkSystemEvents, checkUmp, sendTransaction } from '@acala-network/chopsticks-testing'

import { Network, createContext, createNetworks } from '../../networks'
import { checkEvents } from '../../helpers'

import { acala } from '../../networks/acala'

describe('liquid crowdloan', async () => {
  let chain: Network
  let relaychain: Network
  const crowdloanVault = '132zsjMwGjNaUXF5XjUCDs2cDEq9Qao51TsL9RSUTGZbinVK'
  const palletAccount = '23M5ttkmR6KcoCwfEJnoAgczRQdDq85ZedwGc3uzyPwiCoq8'
  const totalLcdot = 241161597502735335n

  const ctx = createContext()
  const { alice } = ctx

  beforeEach(async () => {
    const { acalaNext, polkadot } = await createNetworks({ acalaNext: undefined, polkadot: undefined }, ctx)
    chain = acalaNext
    relaychain = polkadot

    // assume the crowdloan fund is returned
    await relaychain.dev.setStorage({
      System: {
        Account: [[[crowdloanVault], { providers: 1, data: { free: 24116165n * 10n ** 10n } }]],
      },
    })

    await chain.dev.setStorage({
      Tokens: {
        Accounts: [[[alice.address, acala.lcdot], { free: 1000 * 1e10 }]],
      },
    })

    await relaychain.dev.newBlock({
      unsafeBlockHeight: 17_856_000,
      count: 2,
    } as any)

    return async () => chain.teardown()
  })

  it('works', async () => {
    await sendTransaction(
      chain.api.tx.sudo
        .sudo(
          chain.api.tx.xcmInterface.updateXcmDestWeightAndFee([
            [
              'ProxyReserveTransferAssets',
              {
                refTime: 50e9,
                proofSize: 128 * 1024,
              },
              2e10,
            ],
          ] as any)
        )
        .signAsync(alice)
    )
    await chain.dev.newBlock()

    await sendTransaction(
      chain.api.tx.sudo.sudo(chain.api.tx.liquidCrowdloan.transferFromCrowdloanVault(10e10)).signAsync(alice)
    )
    await chain.dev.newBlock()

    await checkUmp(chain).toMatchSnapshot()

    await relaychain.dev.newBlock()

    await checkSystemEvents(relaychain, 'messageQueue', 'xcmPallet', 'proxy').toMatchSnapshot()

    await chain.dev.newBlock()

    await check(chain.api.query.tokens.accounts(palletAccount, acala.dot)).toMatchSnapshot('pallet account balance')

    const tx0 = await sendTransaction(chain.api.tx.liquidCrowdloan.redeem(9e10).signAsync(alice))

    await chain.dev.newBlock()

    await checkEvents(tx0, 'liquidCrowdloan').toMatchSnapshot('first redeem')

    await sendTransaction(
      chain.api.tx.sudo
        .sudo(chain.api.tx.liquidCrowdloan.transferFromCrowdloanVault(totalLcdot - 10n * 10n ** 10n))
        .signAsync(alice)
    )

    await chain.dev.newBlock()

    await checkUmp(chain).toMatchSnapshot('after all transfer ump')

    await relaychain.dev.newBlock()

    await checkSystemEvents(relaychain, 'messageQueue', 'xcmPallet', 'proxy').toMatchSnapshot(
      'after all transfer relaychain events'
    )

    await chain.dev.newBlock()

    await check(chain.api.query.tokens.accounts(palletAccount, acala.dot)).toMatchSnapshot(
      'pallet account balance after all transfer'
    )

    const tx1 = await sendTransaction(chain.api.tx.liquidCrowdloan.redeem(400e10).signAsync(alice))

    await chain.dev.newBlock()

    await checkEvents(tx1, 'liquidCrowdloan').toMatchSnapshot('second redeem')
  })
})
