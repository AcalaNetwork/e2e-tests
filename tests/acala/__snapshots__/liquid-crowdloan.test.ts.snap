// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`liquid crowdloan > works > after all transfer relaychain events 1`] = `
[
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": 2000,
        },
      },
      "success": false,
      "weightUsed": {
        "proofSize": 0,
        "refTime": 1000000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
  {
    "data": [
      {
        "Complete": {
          "proofSize": 0,
          "refTime": 2000000000,
        },
      },
    ],
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "result": "Ok",
    },
    "method": "ProxyExecuted",
    "section": "proxy",
  },
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": 2000,
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 66000)",
        "refTime": 55000000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`liquid crowdloan > works > after all transfer ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 20000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 20000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "transact": {
          "call": {
            "encoded": "0x1d000059fe89295c2e57d7b4d4d8be9e00a3802e513703ab4b5b424ed0a646e899d3c900630803000100411f03000101006d6f646c6163612f6c71636c000000000000000000000000000000000000000003040000000013e7db26b01ac758030000000000",
          },
          "originType": "SovereignAccount",
          "requireWeightAtMost": 50000000000,
        },
      },
      {
        "refundSurplus": null,
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "parachain": 2000,
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
        },
      },
    ],
  },
]
`;

exports[`liquid crowdloan > works > first redeem 1`] = `
[
  {
    "data": {
      "amount": 90000000000,
      "currencyId": {
        "Token": "DOT",
      },
    },
    "method": "Redeemed",
    "section": "liquidCrowdloan",
  },
]
`;

exports[`liquid crowdloan > works > pallet account balance 1`] = `
{
  "free": 99999216661,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`liquid crowdloan > works > pallet account balance after all transfer 1`] = `
{
  "free": "0x00000000000000000358c71d041ad811",
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`liquid crowdloan > works > second redeem 1`] = `
[
  {
    "data": {
      "amount": 4000000000000,
      "currencyId": {
        "Token": "DOT",
      },
    },
    "method": "Redeemed",
    "section": "liquidCrowdloan",
  },
]
`;

exports[`liquid crowdloan > works > system events 1`] = `
[
  {
    "data": [
      {
        "Complete": {
          "proofSize": 0,
          "refTime": 2000000000,
        },
      },
    ],
    "method": "Attempted",
    "section": "xcmPallet",
  },
  {
    "data": {
      "result": "Ok",
    },
    "method": "ProxyExecuted",
    "section": "proxy",
  },
  {
    "data": {
      "id": "(hash)",
      "origin": {
        "Ump": {
          "Para": 2000,
        },
      },
      "success": true,
      "weightUsed": {
        "proofSize": "(rounded 66000)",
        "refTime": 55000000000,
      },
    },
    "method": "Processed",
    "section": "messageQueue",
  },
]
`;

exports[`liquid crowdloan > works > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 20000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 20000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "transact": {
          "call": {
            "encoded": "0x1d000059fe89295c2e57d7b4d4d8be9e00a3802e513703ab4b5b424ed0a646e899d3c900630803000100411f03000101006d6f646c6163612f6c71636c00000000000000000000000000000000000000000304000000000700e87648170000000000",
          },
          "originType": "SovereignAccount",
          "requireWeightAtMost": 50000000000,
        },
      },
      {
        "refundSurplus": null,
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "parachain": 2000,
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
        },
      },
    ],
  },
]
`;
